### DISCUSSION TITLE
[Camera Support]: Reolink 2-Way Audio on IOS

### DISCUSSION THREAD

**Discussion by NightTrain** [blakeblackshear/frigate#15282] [Camera Support] [answered]
### Describe the problem you are having

I've been testing out a Reolink POE Doorbell. I've currently got video, audio, and 2-way audio working properly in browser over HTTPS. However, I'm unable to get the 2-way audio working via the IOS companion App or via Nabu Casa on any of the IOS browsers that I currently have (Brave, Chrome, Safari). The audio and Video in the IOS companion App or any of the IOS browsers works as expected. 

If I attempt to leverage the microphone via the Frigate Add-on I get a loading indicator that spins forever. Notably the microphone indicator for the phone does turn on. 

If I attempt to leverage the microphone via the custom:frigate card or a custom:webrtc card the microphone indicator comes on, but no audio can be heard from the doorbell itself.

Current Camera Firmware: v3.0.0.4110_2410111120

### Version

0.14.1-f4f3cfa

### What browser(s) are you using?

Chrome 131.0.6778.70 (Official Build) (64-bit)

### Frigate config file

```yaml
mqtt:
  host: <MQTT_HOST_IP>        # YOUR MQTT IP
  port: 1883
  topic_prefix: frigate
  client_id: frigate
  user: <MQTT_USER> # YOUR MQTT USERNAME
  password: <MQTT_PASS>        # YOUT MQTT PASSWORD
  stats_interval: 60

go2rtc:
  streams:
    reolink:
      - ffmpeg:https://<CAMERA_IP>/flv?port=1935&app=bcs&stream=channel0_main.bcs&user=<CAMERA_USER>&<CAMERA_PASS>#video=copy#audio=copy#audio=opus
      - rtsp://<CAMERA_USER>:<CAMERA_PASS>@<CAMERA_IP>:554/h264Preview_01_sub

## THIS PREVENTS THE AUDIO COMING OUT FROM THE DOORBELL SOUNDING MUFFLED
  ffmpeg:
    bin: ffmpeg
    volume: -af "volume=25dB"

##########################################################################

cameras:

  reolink:
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/reolink?video&audio

    detect:
      enabled: false
    record:
      enabled: false
    mqtt:
      enabled: false
    live:
      stream_name: reolink

version: 0.14
```

### Relevant Frigate log output

```shell
Haven't been able to find any relevant logs yet, attempting to sift through debug mode.
```

### Relevant go2rtc log output

```shell
N/A
```

### FFprobe output from your camera

```shell
{
  "producers": [
    {
      "type": "RTSP passive producer",
      "url": "rtsp://127.0.0.1:8554/4340a97b3056398dc4703d78c095f0f7",
      "remote_addr": "127.0.0.1:43082",
      "user_agent": "ffmpeg/go2rtc",
      "sdp": "v=0\r\no=- 0 0 IN IP4 127.0.0.1\r\ns=No Name\r\nc=IN IP4 127.0.0.1\r\nt=0 0\r\na=tool:libavformat LIBAVFORMAT_VERSION\r\nm=video 0 RTP/AVP 96\r\na=rtpmap:96 H264/90000\r\na=fmtp:96 packetization-mode=1; sprop-parameter-sets=Z2QAM6wVFKAoAPGQ,aO48sA==; profile-level-id=640033\r\na=control:streamid=0\r\nm=audio 0 RTP/AVP 97\r\na=rtpmap:97 MPEG4-GENERIC/16000/1\r\na=fmtp:97 profile-level-id=1;mode=AAC-hbr;sizelength=13;indexlength=3;indexdeltalength=3; config=1408\r\na=control:streamid=1\r\nm=audio 0 RTP/AVP 98\r\nb=AS:64\r\na=rtpmap:98 opus/48000/2\r\na=control:streamid=2\r\n",
      "medias": [
        "video, recvonly, H.264 High 5.1",
        "audio, recvonly, MPEG4-GENERIC/16000",
        "audio, recvonly, OPUS/48000/2"
      ],
      "receivers": [
        "96 H264, bytes=2099100608, senders=3",
        "97 MPEG4-GENERIC/16000, bytes=15708743, senders=2",
        "98 OPUS/48000/2, bytes=29946144, senders=2"
      ],
      "recv": 2165085939
    },
    {
      "type": "RTSP active producer",
      "url": "rtsp://<CAMERA_IP>/Preview_01_sub/",
      "remote_addr": "<CAMERA_IP>:554",
      "user_agent": "go2rtc/1.9.2",
      "sdp": "v=0\r\no=- 1732983875475342 1 IN IP4 192.168.0.3\r\ns=Session streamed by \"preview\"\r\nt=0 0\r\na=tool:BC Streaming Media v202210012022.10.01\r\na=type:broadcast\r\na=control:*\r\na=range:npt=now-\r\na=x-qt-text-nam:Session streamed by \"preview\"\r\nm=video 0 RTP/AVP 96\r\nc=IN IP4 0.0.0.0\r\nb=AS:8192\r\na=rtpmap:96 H264/90000\r\na=control:track1\r\nm=audio 0 RTP/AVP 97\r\nc=IN IP4 0.0.0.0\r\nb=AS:8192\r\na=rtpmap:97 MPEG4-GENERIC/16000\r\na=fmtp:97 streamtype=5;profile-level-id=1;mode=AAC-hbr;sizelength=13;indexlength=3;indexdeltalength=3;config=1408\r\na=recvonly\r\na=control:track2\r\nm=audio 0 RTP/AVP 0\r\na=control:track3\r\na=rtpmap:0 PCMU/8000\r\na=sendonly",
      "medias": [
        "video, recvonly, H264",
        "audio, recvonly, MPEG4-GENERIC/16000",
        "audio, sendonly, PCMU/8000"
      ],
      "receivers": [
        "96 H264, bytes=111888828, senders=1",
        "97 MPEG4-GENERIC/16000, bytes=13494282, senders=1"
      ],
      "senders": [
        "0 PCMU/8000, bytes=0, receivers=3"
      ],
      "recv": 127147506
    }
  ],
  "consumers": [
    {
      "type": "RTSP passive consumer",
      "url": "rtsp://127.0.0.1:8554/reolink?video\u0026audio",
      "remote_addr": "127.0.0.1:43072",
      "user_agent": "FFmpeg Frigate/0.14.1-f4f3cfa",
      "sdp": "v=0\r\no=- 1 1 IN IP4 0.0.0.0\r\ns=go2rtc/1.9.2\r\nc=IN IP4 0.0.0.0\r\nt=0 0\r\nm=video 0 RTP/AVP 96\r\na=rtpmap:96 H264/90000\r\na=fmtp:96 packetization-mode=1; sprop-parameter-sets=Z2QAM6wVFKAoAPGQ,aO48sA==; profile-level-id=640033\r\na=control:trackID=0\r\nm=audio 0 RTP/AVP 97\r\na=rtpmap:97 MPEG4-GENERIC/16000\r\na=fmtp:97 profile-level-id=1;mode=AAC-hbr;sizelength=13;indexlength=3;indexdeltalength=3; config=1408\r\na=control:trackID=1\r\n",
      "medias": [
        "video, sendonly, ANY",
        "audio, sendonly, ANY"
      ],
      "senders": [
        "96 H264, bytes=2099100608, receivers=1",
        "97 MPEG4-GENERIC/16000, bytes=15708743, receivers=1"
      ],
      "send": 2138628151
    },
    {
      "type": "WebRTC/WebSocket async passive consumer",
      "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
      "medias": [
        "video, sendonly, VP8, VP9, H264, AV1",
        "audio, sendonly, OPUS/48000/2, G722/8000, PCMU/8000, PCMA/8000, L16, PCML"
      ],
      "senders": [
        "102 H264, bytes=0, receivers=1, overflow=1436785",
        "111 OPUS/48000/2, bytes=0, receivers=1, overflow=195132"
      ]
    },
    {
      "type": "probe",
      "remote_addr": "192.168.1.25:64205",
      "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:132.0) Gecko/20100101 Firefox/132.0",
      "medias": [
        "video, sendonly, ALL",
        "audio, sendonly, ALL",
        "audio, recvonly, ANY"
      ],
      "receivers": [
        "0 ANY, bytes=0, senders=1"
      ],
      "senders": [
        "96 H264, bytes=0, receivers=1",
        "96 H264, bytes=0, receivers=1",
        "97 MPEG4-GENERIC/16000, bytes=0, receivers=1",
        "98 OPUS/48000/2, bytes=0, receivers=1",
        "97 MPEG4-GENERIC/16000, bytes=0, receivers=1"
      ]
    }
  ]
}
```

### Operating system

HassOS

### Install method

HassOS Addon

### Object Detector

CPU (no coral)

### Network connection

Mixed

### Camera make and model

Reolink Video Doorbell POE X003YEKD9X

### Screenshots of the Frigate UI's System metrics pages

![image](https://github.com/user-attachments/assets/b72d2973-25f0-45c6-8cc7-8f6ac1169260)

**Comment by NightTrain** [comment]
An interesting development, I've discovered that when I'm connected to my home wifi network and accessing HA via Nabu Casa and the Brave Browser, I'm able to successfully use 2-way audio, but the same steps over Cellular connection do not work. As a note I'm testing exclusively through the frigate add-on at the moment. 

Frigate logs that I can see when it's working over wifi:
```
2024-12-01 05:24:08.962806253  172.30.32.2 - - [01/Dec/2024:00:24:08 -0500] "GET /live/mse/api/ws?src=reolink HTTP/1.1" 101 5831971 "-" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:24:09.049928562  172.30.32.2 - - [01/Dec/2024:00:24:09 -0500] "GET /api/reolink/latest.webp?h=38&cache=1733030636423 HTTP/1.1" 200 692 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:24:22.952306621  172.30.32.2 - - [01/Dec/2024:00:24:22 -0500] "GET /api/config HTTP/1.1" 200 7791 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:24:22.957191580  172.30.32.2 - - [01/Dec/2024:00:24:22 -0500] "GET /api/profile HTTP/1.1" 200 18 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:24:24.580077352  172.30.32.2 - - [01/Dec/2024:00:24:24 -0500] "GET /api/reolink/latest.webp?h=358&cache=1733030636423 HTTP/1.1" 200 14222 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
```

Frigate logs that I can see when it's working not working over LTE:

```
2024-12-01 05:21:23.471123614  [2024-12-01 00:21:23] urllib3.connectionpool         DEBUG   : Starting new HTTP connection (1): 127.0.0.1:1984
  2024-12-01 05:21:23.471128117  [2024-12-01 00:21:23] urllib3.connectionpool         DEBUG   : http://127.0.0.1:1984 "GET /api/streams?src=reolink&video=all&audio=all&microphone HTTP/11" 200 None
  2024-12-01 05:21:23.472058969  172.30.32.2 - - [01/Dec/2024:00:21:23 -0500] "GET /api/go2rtc/streams/reolink HTTP/1.1" 200 5041 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:21:23.497151304  172.30.32.2 - - [01/Dec/2024:00:21:23 -0500] "GET /api/reolink/latest.webp?h=38&cache=1733030483108 HTTP/1.1" 200 686 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:21:23.564324742  172.30.32.2 - - [01/Dec/2024:00:21:23 -0500] "GET /api/reolink/latest.webp?h=360&cache=1733030483108 HTTP/1.1" 200 14662 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:21:25.113076947  172.30.32.2 - - [01/Dec/2024:00:21:25 -0500] "GET /api/reolink/latest.webp?h=358&cache=1733030483108 HTTP/1.1" 200 14408 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:21:27.376060958  172.30.32.2 - - [01/Dec/2024:00:21:27 -0500] "GET /live/mse/api/ws?src=reolink HTTP/1.1" 101 1711199 "-" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:21:27.751232571  172.30.32.2 - - [01/Dec/2024:00:21:27 -0500] "GET /api/reolink/latest.webp?h=38&cache=1733030483108 HTTP/1.1" 200 682 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:21:27.754178846  172.30.32.2 - - [01/Dec/2024:00:21:27 -0500] "GET /api/reolink/latest.webp?h=239&cache=1733030483108 HTTP/1.1" 200 8228 "<NabuCasa_URL>" "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1" "127.0.0.1, 172.30.32.1"
  2024-12-01 05:21:37.760555235  172.30.32.2 - - [01/Dec/2024:00:21:37 -0500] "GET /ws HTTP/1.1" 101 410637 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:132.0) Gecko/20100101 Firefox/132.0" "192.168.1.25, 172.30.32.1"
```

**Reply by NickM-27** [reply]
it's possible the cellular network is behind a symmetric NAT so webrtc is not able to connect

**Reply by NightTrain** [reply]
@NickM-27 You think that would prevent just the microphone from connecting properly? The audio and video work properly over LTE.

**Reply by NickM-27** [reply]
yes, in most cases webrtc is only used for two way audio in the frigate webUI

**Reply by NightTrain** [reply]
Interesting, is there any way I could prove that via Logs? Any other configuration I could try? Being able to use the 2-way audio when not home is one of my main use cases.

**Reply by NickM-27** [reply]
https://tomchen.github.io/symmetric-nat-test/

**Reply by NightTrain** [reply]
Simple and effective, unfortunately doesn't appear to be my issue. Confirmed my mobile cellular network is a "Normal NAT"

**Reply by NickM-27** [reply]
I just realized you never actually did the webrtc setup https://docs.frigate.video/configuration/live#webrtc-extra-configuration

**Reply by NightTrain** [reply]
I presumed that I didn't need to do that based on the following:
![image](https://github.com/user-attachments/assets/931825b6-5e17-40fc-ab17-dfcad52c26c6)

I also see the following in frigate startup logs:
![image](https://github.com/user-attachments/assets/28404c4a-c794-44a1-aa6d-446140d5d6f8)

However to be sure I've updated my config to the following based on those docs:
```
go2rtc:
  streams:
    reolink:
      - ffmpeg:https://<CAMERA_IP>/flv?port=1935&app=bcs&stream=channel0_main.bcs&user=<CAMERA_USER>&<CAMERA_PASS>#video=copy#audio=copy#audio=opus
      - rtsp://<CAMERA_USER>:<CAMERA_PASS>@<CAMERA_IP>:554/h264Preview_01_sub
    webrtc:
      candidates:
        - <HOST_IP> :8555
        - stun:8555
```

Unfortunately there was no change in the behavior over the cellular network. 

I've also gone ahead and verified that I have the WebRTC configuration turned on in nabu casa:
![image](https://github.com/user-attachments/assets/a54622a1-5ba5-42d7-acc1-2a6417d8fc03)

**Comment by NightTrain** [comment]
An interesting additional note the POE doorbell light does turn on when I activate the microphone over cellular, despite the continual loading indicator. So some amount of information is getting to the doorbell itself before it hangs.

**Comment by NightTrain** [comment]
In an effort to try and narrow down potential places where this could be failing, I set up another Reolink Camera that I had using the same configuration. 

Reolink E1 Pro
Firmware: v3.1.0.3149_2401092478

I ended up encountering the same behavior when attempting to access the 2-way audio via a cellular network. 2-way audio works fine in Desktop Browser and from mobile when connected to the same wifi network. So perhaps this might mean it's something related to my network configuration. I attempted to forward port 8555 to my homeassistant instance, but that didn't seem to have any affect.

**Reply by NickM-27** [reply]
8555 needs to be forwarded to whatever is running Frigate, and it needs to be both udp and tcp

**Reply by NightTrain** [reply]
Correct, I set up the port forwarding for 8555 to go to my home assistant server for both udp and tcp, unfortunately that didn't end up having an affect. I'm trying to figure out if there's any logs I can check to see if there is an attempt to forward. 

I also double checked the frigate logs to make sure the IP matched.

**Reply by NightTrain** [reply]
I don't have much of a network engineering background but I started looking at tcpdump outputs for everything that hits my HA instance. Will follow up on this tomorrow, if there's anything specific I should be looking for please let me know.

**Comment by NightTrain** [comment, marked as answer]
tl;dr answer:

- Stopped using frigate add-on in Home Assistant
- Created an external frigate instance
- Connected with the frigate integration
- Used the configs found here: https://github.com/dermotduffy/frigate-hass-card/issues/1477#issuecomment-2323039575

For anyone that comes across this thread, I was finally able to get this working, albeit not the way I wanted. Originally I was attempting to do this solely with the frigate add-on. Unfortunately after digging through a lot of tcp dumps there was something getting blocked when attempting to connect to webrtc externally with this setup. I switched to a docker setup of frigate and the frigate integration. After setting that up and tweaking my go2rtc settings, I was finally able to get 2-way audio working locally and externally via IOS as long as I was connected via HTTPS.
